cleverhans/attacks/bapp.py

```diff
@@ -0,0 +1,573 @@
+                                                          self.constraint,
+                                                          self.theta)
+
+      elif self.stepsize_search == 'grid_search':
+        # Grid search for stepsize.
+        epsilons = np.logspace(-4, 0, num=20, endpoint=True) * dist
+        epsilons_shape = [20] + len(self.shape) * [1]
+        perturbeds = perturbed + epsilons.reshape(epsilons_shape) * update
+        idx_perturbed = decision_function(perturbeds)
+
+        if np.sum(idx_perturbed) > 0:
+          # Select the perturbation that yields the minimum distance # after binary search.
+          perturbed, dist_post_update = binary_search_batch(sample,
+                                                            perturbeds[idx_perturbed], decision_function, self.shape,
+                                                            self.constraint, self.theta)
+
+      # compute new distance.
+      dist = compute_distance(perturbed, sample, self.constraint)
+      if self.verbose:
+        print('iteration: {:d}, {:s} distance {:.4E}'.format(
+            j+1, self.constraint, dist))
+
+    perturbed = np.expand_dims(perturbed, 0)
+    return perturbed
+
+
+def clip_image(image, clip_min, clip_max):
+  """ Clip an image, or an image batch, with upper and lower threshold. """
+  return np.minimum(np.maximum(clip_min, image), clip_max)
+
+
+def compute_distance(x_ori, x_pert, constraint='l2'):
+  """ Compute the distance between two images. """
+  if constraint == 'l2':
+    dist = np.linalg.norm(x_ori - x_pert)
+  elif constraint == 'linf':
+    dist = np.max(abs(x_ori - x_pert))
+  return dist
+
+def approximate_gradient(decision_function, sample, num_evals,
+                         delta, constraint, shape, clip_min, clip_max):
+  """ Gradient direction estimation """
+  # Generate random vectors.
+  noise_shape = [num_evals] + list(shape)
+  if constraint == 'l2':
+    rv = np.random.randn(*noise_shape)
+  elif constraint == 'linf':
+    rv = np.random.uniform(low=-1, high=1, size=noise_shape)
+
+  axis = tuple(range(1, 1 + len(shape)))
+  rv = rv / np.sqrt(np.sum(rv ** 2, axis=axis, keepdims=True))
+  perturbed = sample + delta * rv
+  perturbed = clip_image(perturbed, clip_min, clip_max)
+  rv = (perturbed - sample) / delta
+
+  # query the model.
+  decisions = decision_function(perturbed)
+  decision_shape = [len(decisions)] + [1] * len(shape)
+  fval = 2 * decisions.astype(np_dtype).reshape(decision_shape) - 1.0
+
+  # Baseline subtraction (when fval differs)
+  if np.mean(fval) == 1.0:  # label changes.
+    gradf = np.mean(rv, axis=0)
+  elif np.mean(fval) == -1.0:  # label not change.
+    gradf = - np.mean(rv, axis=0)
+  else:
+    fval = fval - np.mean(fval)
+    gradf = np.mean(fval * rv, axis=0)
+
+  # Get the gradient direction.
+  gradf = gradf / np.linalg.norm(gradf)
+
+  return gradf
+
+
+def project(original_image, perturbed_images, alphas, shape, constraint):
```

Comment:
You can probably reuse https://github.com/tensorflow/cleverhans/blob/f70ca7e000dadd6ace5aeff15bba0e960e8c1384/cleverhans/utils_tf.py#L358

Reply:
'clip_eta' is not applicable here. 'clip_eta' with ord=2 first calculates the norm of a tensor, and then divides the tensor by the norm. Thus clip_eta is equivalent to projection onto the unit sphere centered at the origin. 'project' projects a tensor to a sphere centered at a given point (not necessarily origin) with a given radius (not necessarily its norm). 'project' seems to be a more general version of 'clip_eta'.